lib/blockchain.go

```diff
@@ -487,14 +490,27 @@ type Blockchain struct {
 	timer *Timer
 }
 
+func (bc *Blockchain) addNewBlockNodeToBlockIndex(blockNode *BlockNode) {
+	bc.blockIndexByHash[*blockNode.Hash] = blockNode
+	bc.blockIndexByHeight[uint64(blockNode.Height)] = append(bc.blockIndexByHeight[uint64(blockNode.Height)], blockNode)
```

Comment:
I don't love that the two maps can now go out of sync. IMO, the two maps should contain an identical set of nodes, and should be resilient to edge cases where the caller make a copy of an existing blockNode... or the caller calls addNewBlockNodeToBlockIndex more than once per node.

What do you think about updating the blockIndexByHeight to instead be of type `map[uint64]map[BlockHash]*BlockNode` instead, where it's a nested map that index blocks by height first and then by block hash. This way, you can guarantee that both the blockIndexByHash and blockIndexByHeight maps will contain an identical set of elements regardless of how addNewBlockNodeToBlockIndex is used (or misused) by the caller. Here's 

```suggestion
	bc.blockIndexByHash[*blockNode.Hash] = blockNode
	if _, exists := bc.blockIndexByHeight[uint64(blockNode.Height)]; !exists {
		bc.blockIndexByHeight[uint64(blockNode.Height)] = map[BlockHash]*BlockNode{}
	}
	bc.blockIndexByHeight[uint64(blockNode.Height)][*blockNode.Hash] = blockNode
```

And then, when you later need to fetch all blockNodes by height, you can define a separate helper function that uses the new map:
```
func (bc *Blockchain) getAllBlockNodesByHeight(uint64 blockHeight) []*BlockNode {
	return collections.MapValues(bc.blockIndexByHeight[blockHeight])
}
```